RFC Draft.md

```diff
@@ -382,9 +382,249 @@ const uploadedObjectReference = await uploadTask.result;
+import { GraphQLQuery, GraphQLSubscription } from '@aws-amplify/api';
+import { createTodo } from './graphql/mutations';
+import { onCreateTodo } from './graphql/subscriptions';
+import {
+  CreateTodoInput,
+  CreateTodoMutation,
+  OnCreateTodoSubscription,
+} from './API';
+// so many imports from disparate modules 
+
+function createMutation() {
+  const createInput: CreateTodoInput = {
+    name: 'Improve API TS support ',
+  };
+
+  const res = await API.mutate(createTodo, {
+      input: createInput,
+  });
+
+  // The returned data is the result of the request. If there are more than one queries/mutations in a request,
+  // then the return value stays the same as v5. i.e. res.createTodo.data
+  const newTodo = res;
+
+  console.log('new todo created', newTodo);
+}
+
+function subscribeToCreate() {
+  const sub = API.subscribe(onCreateTodo).on({
+    next: (message) => {
+      // Return value immediately available instead of `value?.data?.onCreateTodo`
+      const newTodo = message;
+      console.log('new todo created', newTodo);
+    }
+  });
+}
+```
+
+## Type safety for GraphQL query, mutation, subscription inputs
+
+In v6, we want to ensure type safety on GraphQL inputs if you use one of the generated GraphQL query, mutation, or subscription.
+
+**Current Usage (v5)**
+```ts
+import { updateTodo } from './graphql/mutations';
+import { CreateTodoInput } from './API';
+
+const createInput: CreateTodoInput = {
+  name: todoName,
+  description,
+};
+
+const res = await API.graphql<GraphQLQuery<UpdateTodoMutation>>(
+  graphqlOperation(updateTodo, {
+    // passing an object of type CreateTodoInput (that's missing
+    // a required field for updates `id`) into an update mutation's input
+    // does not surface a type error. This will only throw a runtime error after
+    // the mutation request gets rejected by AppSync
+    input: createInput,
+  })
+);
+```
+
+**Proposed Usage (v6)**
+```ts
+import { updateTodo } from './graphql/mutations';
+import { CreateTodoInput } from './API';
+
+const createInput: CreateTodoInput = {
+  name: todoName,
+  description,
+};
+
+const res = await API.graphql<GraphQLQuery<UpdateTodoMutation>>(
+  graphqlOperation(updateTodo, {
+    input: createInput,
+    //^? Throws a type error. `input` must be of type `UpdateTodoInput`
+  })
+);
+```
+
+
+
+## [Bug fix: Remove `any` cast needed for subscriptions](https://github.com/aws-amplify/amplify-js/issues/7589#issuecomment-1258596131)
+
+In v5, there's a type mismatch bug for GraphQL subscriptions that forces the developer to cast to `any` to unsubscribe. We plan on fixing this for v6.
+
+**Current Usage (v5)**
+```ts
+  subscription.value = (API.graphql(graphqlOperation(
+    SUBSCRIBE_USER,
+    { id: userId }
+  )) as any).subscribe({ next: onSubscribe })
+// . . .
+  (subscription.value as any).unsubscribe()
+```
+
+**Proposed Usage (v6)**
+```ts
+  subscription = API.subscribe(
+    SUBSCRIBE_USER,
```

Comment:
Where does `SUBSCRIBE_USER` come from?

Per our discussion around V6 goals, we can only automatically type this if this is one of the code generated queries. (Or if the customer creates a "nominally typed" query themselves.)

Reply:
added imports